src/LagrangianParticleTracking/calculate_particle_tendency_contributions.jl

```diff
@@ -0,0 +1,75 @@
+        tendency = getproperty(model.timestepper.Gⁿ, field.tracer)
+
+        Gp_kernel! = calculate_particle_tendency_kernel!(device(arch), workgroup, worksize)
+        Gp_event = Gp_kernel!(property, tendency, particles, model.tracers, model.grid)
+
+        push!(events, Gp_event)
+    end
+
+    wait(device(arch), MultiEvent(Tuple(events)))
+
+    return nothing
+end
+
+calculate_particle_tendency_contributions!(model) = calculate_particle_tendency_contributions!(model.particles, model)
+calculate_particle_tendency_contributions!(particles::Nothing, model) = nothing
+calculate_particle_tendency_contributions!(particles, model) = !isnothing(particles.parameters) ? (:active_properties in keys(particles.parameters) ? calculate_particle_tendency_contributions!(particles, model, particles.parameters.active_properties) : nothing) : nothing
+calculate_particle_tendency_contributions!(particles, model, active_fields::Nothing) = nothing
+
+@inline function get_mask(x, y, z, grid::AbstractGrid{FT, TX, TY, TZ}, loc) where {FT, TX, TY, TZ}
+    (fi, i::Int), (fj, j::Int), (fk, k::Int) = modf.(fractional_indices(x, y, z, loc, grid))
+
+    nodesi = arch_array(grid.architecture, zeros(Int, 2, 2, 2))
+    nodesj = arch_array(grid.architecture, zeros(Int, 2, 2, 2))
+    nodesk = arch_array(grid.architecture, zeros(Int, 2, 2, 2))
+    d = arch_array(grid.architecture, zeros(2, 2, 2))
+    
+    if (fi, fj, fk) == (0, 0, 0) #particle on grid point special case
+        @inbounds nodesi[1, 1, 1], nodesj[1, 1, 1], nodesk[1, 1, 1], d = i, j, k, 1
+    else
+        for (a, di) in enumerate([fi, 1-fi]), (b, dj) in enumerate([fj, 1-fj]), (c, dk) in enumerate([fk, 1-fk])
+            @inbounds begin
+                d[a,b,c] = sqrt(di^2+dj^2+dk^2)
+                nodesi[a, b, c] = getnode(TX(), i+a, grid.Nx)
+                nodesj[a, b, c] = getnode(TY(), j+b, grid.Ny)
+                nodesk[a, b, c] = getnode(TZ(), k+c, grid.Nz)
+            end
+        end
+        
+        normfactor=1/sum(1 ./ d)
+        d ./= normfactor
+    end
+
+    return nodesi, nodesj, nodesk, d
+end
+
+@inline getnode(::Bounded, i, N) = min(max(i, 1), N)
+@inline getnode(::Periodic, i, N) = i < 1 ? N - i : (i > N ? 1 + i : i)
+
+@kernel function calculate_particle_tendency_kernel!(property, tendency, particles, model, grid)
+    p = @index(Global)
+    density = :density in keys(particles.parameters) ? particles.parameters.density : 1
+    
+    LX, LY, LZ = location(tendency)
+nodesi, nodesj, nodesk, d = @inbounds get_mask(particles.properties.x[p], particles.properties.y[p], particles.properties.z[p], grid, (LX(), LY(), LZ()))
+        
+    for (idx, i) in enumerate(nodesi) #tried a vector version of this and was much slower
+        @inbounds begin
+            j = nodesj[idx]; k = nodesk[idx]
+            vol = volume(i, j, k, grid, LX(), LY(), LZ())
+            value = density * property[p] / (d[idx] * vol)
+
+            tendency[i, j, k] += value
+        end
+    end
```

Comment:
Corresponding to changes [here](https://github.com/CliMA/Oceananigans.jl/pull/2724/files#r964376612)

```suggestion
    nodes, normfactor = @inbounds get_mask(particles.properties.x[p], particles.properties.y[p], particles.properties.z[p], grid, (LX(), LY(), LZ()))
        
    @unroll for (i, j, k, d) in nodes #tried a vector version of this and was much slower
        i, j, k = (getnode(TX(), i, grid.Nx), getnode(TY(), j, grid.Ny), getnode(TZ(), k, grid.Nz))
        vol = volume(i, j, k, grid, LX(), LY(), LZ())
        value = density * @inbounds property[p] * normfactor/ (d * vol)
        @inbounds tendency[i, j, k] += value
    end
```